resources/custom-resources/app-services-macro/src/main/java/com/amazon/aws/partners/saasfactory/saasboost/ApplicationServicesMacro.java

```diff
@@ -52,8 +55,146 @@ public Map<String, Object> handleRequest(Map<String, Object> event, Context cont
+            return response;
+        }
+        
+        String extensionsError = updateTemplateForPooledExtensions(templateParameters, template);
+        if (extensionsError != null) {
+            response.put(ERROR_MSG, extensionsError);
+            return response;
+        }
+
+        response.put(FRAGMENT, template);
+        response.put(STATUS, SUCCESS);
+        return response;
+    }
+
+    protected static String updateTemplateForPooledExtensions(
+            final Map<String, Object> templateParameters,
+            Map<String, Object> template) {
+        Set<String> processedExtensions = new HashSet<String>();
+        if (templateParameters.containsKey("AppExtensions") && template.containsKey("Resources")) {
+            String applicationExtensions = (String) templateParameters.get("AppExtensions");
+            if (Utils.isNotEmpty(applicationExtensions)) {
+                String[] extensions = applicationExtensions.split(",");
+                for (String extension : extensions) {
+                    if (processedExtensions.contains(extension)) {
+                        LOGGER.warn("Skipping duplicate extension {}", extension);
+                        continue;
+                    }
+                    switch (extension) {
+                        case "s3": {
+                            String s3Error = updateTemplateForS3(templateParameters, template);
+                            if (s3Error != null) {
+                                LOGGER.error("Processing S3 extension failed: {}", s3Error);
+                                return s3Error;
+                            }
+                            break;
+                        }
+                        default: {
+                            LOGGER.warn("Skipping unknown extension {}", extension);
+                        }
+                    }
+                    processedExtensions.add(extension);
+                }
+            } else {
+                LOGGER.debug("Empty AppExtensions parameter, skipping updating template for extensions");
+            }
+        } else {
+            LOGGER.error("Invalid template, missing AppExtensions parameter or missing Resources");
+            return "Invalid template, missing AppExtensions parameter or missing Resources";
+        }
+        return null;
+    }
+
+    protected static String updateTemplateForS3(final Map<String, Object> templateParameters,
+            Map<String, Object> template) {
+        if (templateParameters.containsKey("Environment")
+                && templateParameters.containsKey("LoggingBucket")) {
+            String environmentName = (String) templateParameters.get("Environment");
+            // Bucket Resource
+            Map<String, Object> s3Resource = s3Resource(environmentName,
+                    (String) templateParameters.get("LoggingBucket"));
+            ((Map<String, Object>) template.get("Resources")).put("S3Extension", s3Resource);
```

Comment:
Prefer we call this Resource `storage` or `objects` (lowercase) instead of `S3Extension` to try and match a little better the pattern we use for other provisioned S3 buckets and to make its use maybe a little more clear.